### PR TITLE
Check flags for `newMethod`, `newVal` and `newBind`

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -4229,6 +4229,13 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
       /** Is this symbol `abstract` */
       def Abstract: Flags
 
+      /** Is this symbol is `abstract override`
+       *
+       *  A trait that has only abstract methods as members
+       *  and therefore can be represented by a Java interface.
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:4235]: That description looks incorrect `abstract override` is a specific concept in Scala that is unrelated to Java interfaces: https://www.scala-lang.org/files/archive/spec/2.13/05-classes-and-objects.html#abstract-override

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -2484,13 +2486,28 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       def newMethod(owner: Symbol, name: String, tpe: TypeRepr): Symbol =
         newMethod(owner, name, tpe, Flags.EmptyFlags, noSymbol)
       def newMethod(owner: Symbol, name: String, tpe: TypeRepr, flags: Flags, privateWithin: Symbol): Symbol =
+        import Flags.*
+        // TODO: missing AbsOverride
+        checkValidFlags(flags.toTermFlags, Private | Protected | Override | Deferred | Final | Method | Implicit | Given | Local | JavaStatic) // Synthetic | ExtensionMethod | Exported | Erased | Infix | Invisible
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2491]: It would be nicer to create a val for each of these long lists of flags. Also couldn't this be checked in -Ycheck?

**Review by smarter — CHANGES_REQUESTED** [review]

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2491]: The idea of the `-Xcheck-macros` flag is to help users find the location in their implementation where the bug started. Additionally, waiting until `-Ycheck` can result in other derived failures or crashes before we reach the checks.

File: library/src/scala/quoted/Quotes.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:4235]: I aligned the doc with https://github.com/lampepfl/dotty/blob/main/compiler/src/dotty/tools/dotc/core/Flags.scala#L282-L288

Comment (smarter) [reply library/src/scala/quoted/Quotes.scala:4235]: This part of the documentation comment is for PureInterface, not for AbsOverride

Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: I updated the docs.

Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: I noticed that `AbsOverride` is pickled as `Abstract | Override` and then unpickled as `AbsOverride`. Maybe we should not have this flag in the reflection interface and handle `AbsOverride` as if both `Abstract` and `Override`.

Comment (smarter) [reply library/src/scala/quoted/Quotes.scala]: > I noticed that AbsOverride is pickled as Abstract | Override and then unpickled as AbsOverride. Maybe we should not have this flag in the reflection interface and handle AbsOverride as if both Abstract and Override.

I'm not sure diverging from the compiler implementation is a good idea here. In the specification, `abstract override` is a separate concept from `abstract` and `override`, it just reuses existing keywords. The tasty serialization format also reuses the flags bits, but that's just an optimization.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 3]
```diff
@@ -2842,6 +2854,10 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       def Synthetic: Flags = dotc.core.Flags.Synthetic
       def Trait: Flags = dotc.core.Flags.Trait
       def Transparent: Flags = dotc.core.Flags.Transparent
+
+      private[QuotesImpl] def validMethodFlags: Flags = Private | Protected | Override | Deferred | Final | Method | Implicit | Given | Local | JavaStatic | AbsOverride // Synthetic | ExtensionMethod | Exported | Erased | Infix | Invisible
+      private[QuotesImpl] def validValFlags: Flags = Private | Protected | Override | Deferred | Final | Param | Implicit | Lazy | Mutable | Local | ParamAccessor | Module | Package | Case | CaseAccessor | Given | Enum | JavaStatic | AbsOverride //  Synthetic | Erased | Invisible
+      private[QuotesImpl] def validBindFlags: Flags = Case // | Implicit | Given | Erased
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2860]: What are the commented-out flags meant to represent? A comment might be helpful.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 4]
```diff
@@ -2842,6 +2854,10 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
       def Synthetic: Flags = dotc.core.Flags.Synthetic
       def Trait: Flags = dotc.core.Flags.Trait
       def Transparent: Flags = dotc.core.Flags.Transparent
+
+      private[QuotesImpl] def validMethodFlags: Flags = Private | Protected | Override | Deferred | Final | Method | Implicit | Given | Local | JavaStatic | AbsOverride // Synthetic | ExtensionMethod | Exported | Erased | Infix | Invisible
```
Comment (smarter) [inline compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala:2858]: - It looks like the list here is duplicated in the documentation of newMethod, we should add a comment here to remember to update the list in the documentation.
- I don't think users can create JavaStatic methods.

File: library/src/scala/quoted/Quotes.scala [hunk 5]
```diff
@@ -3816,7 +3816,7 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
       *
       *  @param parent The owner of the binding
       *  @param name The name of the binding
-      *  @param flags extra flags to with which the symbol should be constructed
+      *  @param flags extra flags to with which the symbol should be constructed. `Case` flag will be added. Can be `Case`
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3819]: If Case is always added and Case is the only valid flag, it sounds like this parameter is not actually necessary?

File: library/src/scala/quoted/Quotes.scala [hunk 6]
```diff
@@ -4373,6 +4373,18 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+      /** Is this symbol is labeled with of abstract & override
+       *
+       *  The override modifier has an additional significance when combined with the abstract modifier.
+       *  That modifier combination is only allowed for value members of traits.
+       *
+       *  We call a member MM of a template incomplete if it is either abstract (i.e. defined by a declaration), or it is labeled abstract and override and every member overridden by MM is again incomplete.
+       *
+       *  Note that the abstract override modifier combination does not influence the concept whether a member is concrete or abstract.
+       *  A member is abstract if only a declaration is given for it; it is concrete if a full definition is given.
+       */
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:4385]: I've rewritten the documentation to be more user-friendly
```suggestion
      /** Is this an abstract override method?
       *
       *  This corresponds to a definition declared as "abstract override def" in the source.
       * See https://stackoverflow.com/questions/23645172/why-is-abstract-override-required-not-override-alone-in-subtrait for examples.
       */
```

**Review by smarter — CHANGES_REQUESTED** [review]

File: library/src/scala/quoted/Quotes.scala (hunk 5 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3819]: In the future, we might have more flags (see `validBindFlags`). 

We could add an overload of `newBind` that does not require the flags.
